common/src/leap/soledad/common/couch.py

```diff
@@ -476,9 +434,14 @@ def __init__(self, url, dbname, replica_uid=None, ensure_ddocs=True):
             self._set_replica_uid(replica_uid)
         if ensure_ddocs:
             self.ensure_ddocs_on_db()
-        # initialize a thread pool for parallelizing get_docs()
-        self._sem_pool = threading.BoundedSemaphore(
-            value=self.MAX_GET_DOCS_THREADS)
+        self._cache = None
```

Comment:
Can't agree more. For me it also doesn't make sense to name this class CouchDatabase. It has a self._database attribute that makes logic even more confusing. We can try to isolate leap_mx usage of Soledad into a function, making underlying implementation free.

Also, couch.py has 1.5k lines and is awful to read, we should take the CouchDB code out of Backend logic and try to centralize it on a separated module. This refactor should make it easy to change CouchDB usage. SoledadBackend would be a CommonBackend subclass (today's CouchDatabase) that receives an Database on constructor, implementing needed storage methods. This design would allow us to go for any library, fix bugs faster or even provide an alternative database implementation without modifying current backend logic, which is already fine and well tested. Even mocking the database during backend tests would be easier with a mocked "Database" instance. Caching could be applied on a subclass, something like CachedCouchDatabase, providing the same operations with cache.

I didn't started this together on that PR because current change is already too big and I believe that small steps are the best way to review and discuss. :)
(and yes, I didn't want to mess up with code that could break leap_mx or show unknown side effects. That would be too risky with the current already big change)
Feel free to call me on mumble for further discussions on that! :+1: